types/recommendationrequest.go

```diff
@@ -0,0 +1,64 @@
+/*
+Copyright 2019 The MayaData Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package types
+
+import (
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// CStorPoolClusterRecommendationRequest is a kubernetes custom
+// resource that represents configuration input or requirement
+// to create or scale one CStor pool.
+type CStorPoolClusterRecommendationRequest struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	Spec CStorPoolClusterRecommendationRequestSpec `json:"spec"`
+}
+
+// CStorPoolClusterRecommendationRequestSpec contains input or requirement
+// to create or scale one CStorpoolcluster.
+type CStorPoolClusterRecommendationRequestSpec struct {
+	// Name represents name of the cspc.
+	Name string `json:"name"`
+	// Namespace represents namespace of the cspc.
+	Namespace string `json:"namespace"`
+	// PoolCapacity represents requested capacity for one pool
+	PoolCapacity resource.Quantity `json:"capacity"`
+	// DataConfig represents raid configuration for data devices.
+	DataConfig RaidGroupConfig `json:"dataConfig"`
+	// WriteCacheConfig represents raid configuration for write cache devices.
+	// If this field is nil then write cache is disabled.
+	WriteCacheConfig *RaidGroupConfig `json:"writeCacheConfig"`
+	// ReadCacheConfig represents raid configuration for read cache devices.
+	// If this field is nil then read cache is disabled.
+	ReadCacheConfig *RaidGroupConfig `json:"readCacheConfig"`
+}
+
+// RaidGroupConfig contains raid type and device(s)
+// count for a raid group
+type RaidGroupConfig struct {
+	// Type is the raid group type
+	// Supported values are : stripe, mirror, raidz and raidz2
+	Type string `json:"type"`
+	// -- stripe DeviceCount = 1
+	// -- mirror DeviceCount = 2
+	// -- raidz DeviceCount = (2^n + 1) default is (2 + 1)
+	// -- raidz2 DeviceCount = (2^n + 2) default is (4 + 2)
+	DeviceCount int64 `json:"deviceCount"`
```

Comment:
IMO this should be removed. Let us come up with different terminology that all converge to one field i.e. `RAIDGroupConfig.Type`

Reply:
Since no-of devices in a group can have different values, we need some identifier.
Eg. Raidz1 can have 2+1 or 4+1, both are valid.

Reply:
Yes. I agree. If there is no standard that clearly points a 4+1 RAIDZ1 e.g. `RAIDZ41` then let us have this as optional field that defaults to 2+1 in this case.

Reply:
Hmm.. Raidz41 is not a valid type.
Agree this is an optional field only for Raidz1 and Raidz2